### ISSUE TITLE
Enhancement(utils): Update types ESLint v9 Linter class (flat config)

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#8968]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Relevant Package

utils

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

Splitting out of #8211 / https://github.com/typescript-eslint/typescript-eslint/issues/8211#issuecomment-2044094079:

> ## API Users (mainly those that use `utils/ts-eslint`)
> * Switch Linter to flat config by default
> 
> We'd need to look more into the difference between the before and after here as to whether it actually makes a difference for consumers of our package. If it doesn't then all good we can tick and move on! If it does then we need to figure out a path forward here with the types. EG how can we support types that depend on the version of `eslint` the user has installed?

In summary: we need to update the `TSESLint.Linter` class within `@typescript-eslint/utils/ts-eslint` to have the same default behavior as the core `ESLint.Linter` class.

### Additional Info

This isn't a `good first issue`. Someone who's already familiar with the classes & utilities in question should tackle this.

💖

**Comment by JoshuaKGoldberg** [comment]
Hmm, looking at https://github.com/eslint/eslint/pull/17851, I think we get this for free by virtue of our `Linter` being a re-export of the ESLint class?

https://github.com/typescript-eslint/typescript-eslint/blob/fa679550052f122c82bc4fd88d2e2c9138824bc3/packages/utils/src/ts-eslint/Linter.ts#L308-L310

cc @bradzacher - I think we can mark this as already done?

**Comment by bradzacher** [comment]
Specifically it's the types we need to correct here. Make sure that the types accept a flat config.

**Comment by JoshuaKGoldberg** [comment]
Right - and they do, I think?

https://github.com/typescript-eslint/typescript-eslint/blob/fa679550052f122c82bc4fd88d2e2c9138824bc3/packages/utils/src/ts-eslint/Linter.ts#L80-L84

https://github.com/typescript-eslint/typescript-eslint/blob/fa679550052f122c82bc4fd88d2e2c9138824bc3/packages/utils/src/ts-eslint/Linter.ts#L137

**Comment by bradzacher** [comment]
https://eslint.org/docs/latest/use/migrate-to-9.0.0#-linter-now-expects-flat-config-format

> If you still need the v8.x Linter functionality, pass configType: "eslintrc" to the constructor

We don't support that flag in the constructor options I think.

We might want to generically parameterise it as well? Maybe? Or maybe we don't care and just the constructor option is fine.

**Comment by JoshuaKGoldberg** [comment]
Yeah I would really like to not have to parameterize it. 😬 Quite a lot of work for very little gain. If someone files an issue I think we can always see if they have a use case that really benefits from it.

**Comment by JoshuaKGoldberg** [comment]
#8999 was merged into `v8`. ✅
